### PR TITLE
[WIP] point cloud renderer optimizations

### PR DESCRIPTION
**Pull request by ds58**

### CODE REVIEW

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java [hunk 1]
```diff
@@ -200,22 +193,29 @@ public void updateMesh(RawImage depthImage)
...
          vertices = new float[floatsPerVertex * pixelCount];
 
+         for (int i = 0; i < pixelCount; i++)
+         {
+            int offset = i * floatsPerVertex;
+            int x = i % width;
+            int y = i / width;
+            vertices[offset + 1] = x;
+            vertices[offset + 2] = y;
+         }
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java]: Oh yeah, this is great

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java [hunk 2]
```diff
@@ -266,18 +266,25 @@ public void updateMesh(RawImage depthImage, RawImage colorImage)
...
       if (vertices.length != floatsPerVertex * points.size())
+      {
          vertices = new float[floatsPerVertex * points.size()];
 
+         for (int i = 0; i < points.size(); i++) {
+            Point3DReadOnly point = points.get(i);
+            int offset = i * floatsPerVertex;
+            vertices[offset + 1] = point.getY32();
+            vertices[offset + 2] = point.getZ32();
+         }
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java]: This wouldn't work the same :(

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java [hunk 3]
```diff
@@ -266,18 +266,25 @@ public void updateMesh(RawImage depthImage, RawImage colorImage)
...
+
       // Copy points over
       IntStream.range(0, points.size()).parallel().unordered().forEach(i ->
       {
          Point3DReadOnly point = points.get(i);
          int offset = i * floatsPerVertex;
-
          vertices[offset] = point.getX32();
-         vertices[offset + 1] = point.getY32();
-         vertices[offset + 2] = point.getZ32();
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java]: This needs to be undone

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java [hunk 4]
```diff
@@ -295,10 +302,7 @@ public void updateMesh(Point3DReadOnly[] points)
       {
          Point3DReadOnly point = points[i];
          int offset = i * floatsPerVertex;
-
          vertices[offset] = point.getX32();
-         vertices[offset + 1] = point.getY32();
-         vertices[offset + 2] = point.getZ32();
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java]: This too

**Review by TomaszTB — CHANGES_REQUESTED** [review]
Great change for the depth image stuff! But we need to copy over the entire point cloud each time :( I'm not too worried about optimizing that one, since it's just Java to native memory copy (not the native -> Java -> native as with the depth images)

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java [hunk 5]
```diff
@@ -1,12 +1,7 @@
 package us.ihmc.rdx.ui.graphics;
 
-import com.badlogic.gdx.graphics.Color;
-import com.badlogic.gdx.graphics.Mesh;
-import com.badlogic.gdx.graphics.Pixmap;
+import com.badlogic.gdx.graphics.*;
```
Comment (TomaszTB) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.java]: Off topic, but I wonder if your formatter is different than mine? Mine didn't collapse the import into a *

**Review by TomaszTB — APPROVED** [review]

File: ihmc-high-level-behaviors/src/libgdx/resources/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.glsl [hunk 6]
```diff
@@ -77,20 +77,20 @@ void main()
...
-   float depthInMeters = a_position.x * u_depthDiscretization;
+   float depthInMeters = a_depthData.x * u_depthDiscretization;
 
    if (depthInMeters == 0.0f)
       return;
 
-   uint x = uint(a_position.y);
-   uint y = uint(a_position.z);
+   uint x = gl_VertexID % 1280;
+   uint y = gl_VertexID / 1280;
```
Comment (ds58) [inline ihmc-high-level-behaviors/src/libgdx/resources/us/ihmc/rdx/ui/graphics/RDXRawImagePointCloudRenderer.glsl:97]: Need to remove the resolution width here

**Pull request merged**